controllers/ingressnodefirewall_controller.go

```diff
@@ -271,6 +276,39 @@ func (r *IngressNodeFirewallReconciler) buildNodeStates(
 			return nil, err
 		}
 
+		// examine DeletionTimestamp to determine if object is under deletion
+		if firewallObj.ObjectMeta.DeletionTimestamp.IsZero() {
+			// The object is not being deleted, so if it does not have our finalizer,
+			// then lets add the finalizer and update the object. This is equivalent
+			// registering our finalizer.
+			if !stringInArray(ingressNodeFirewallFinalizerName, firewallObj.ObjectMeta.Finalizers) {
+				firewallObj.ObjectMeta.Finalizers = append(firewallObj.ObjectMeta.Finalizers, ingressNodeFirewallFinalizerName)
```

Comment:
I think that you are mutating the object, you have to create a deepcopy with `firewallObjClone := firewallObj.DeepCopy` 

that may explain the races if you are mutating the cache objects

Reply:
oh, this is kubebuilder, I don't know how it implements the Lists

Reply:
but it seems the `infList` is passed as parameter, you should DeepCopy to avoid mutating the list if someon else is using it